File: Sources/ActionRingCore/Controller/RingController.swift
import Foundation
#if canImport(AppKit) && canImport(SwiftUI)
import AppKit
import SwiftUI

/// The conductor. Owns every subsystem and wires the core loop:
///
/// `trigger down` → (unless fullscreen) show ring at cursor → `drag` highlights
/// the angular slot → `trigger up` fires the highlighted slot's action (or
/// cancels in the dead zone).
@available(macOS 14.0, *)
@MainActor
public final class RingController {

    public let profileStore: ProfileStore
    public let appDetector: AppDetector
    public let contextEngine: ContextEngine
    public let fullscreenDetector: FullscreenDetector
    public let eventTap: EventTapManager
    private let executor: ActionExecutor

    private let model: RingViewModel
    private var window: RingWindow?

    /// True while the ring is on screen for the current press.
    /// Ring interaction state. Supports two natural gestures:
    ///  • **Hold**: press-and-hold → drag to a slot → release to fire.
    ///  • **Click**: a quick tap opens a *sticky* ring that stays up; move to a
    ///    slot and click again to fire (click in the center to cancel).
    private enum RingState { case closed, holding, sticky }
    private var state: RingState = .closed
    /// When the current ring opened (to tell a quick tap from a deliberate hold).
    private var openedAt = Date.distantPast
    /// A tap shorter than this with no slot selected opens the sticky ring.
    private let tapThreshold: TimeInterval = 0.30

    /// Set when a trigger-down was suppressed, so the matching trigger-up is also
    /// suppressed — even if the poll closed the ring first (which would otherwise
    /// let a dangling up reach the focused app).
    private var expectingTriggerUp = false
    /// Drives cursor tracking + release detection while the ring is open.
    private var pollTimer: Timer?
    /// In preview mode the ring follows the cursor but ignores button state and
    /// never executes; it auto-dismisses on a (cancellable) timer.
    private var previewMode = false
    private var previewDismiss: DispatchWorkItem?
    /// Safety auto-cancel for a sticky ring left open.
    private var stickyTimeout: DispatchWorkItem?

    /// Last action result, surfaced for diagnostics/UI if desired.
    public private(set) var lastResult: ActionExecutorResult?
    public var onActionExecuted: ((RingSlot, ActionExecutorResult) -> Void)?

    /// True only when the event tap is actually installed and listening — i.e.
    /// the trigger button will open the ring right now. This is stronger than
    /// "accessibility granted": the tap can still need a (re)arm after a grant.
    public var isArmed: Bool { eventTap.isEnabled }

    public init(profileStore: ProfileStore = ProfileStore()) {
        self.profileStore = profileStore
        self.appDetector = AppDetector()
        self.contextEngine = ContextEngine(appDetector: appDetector, provider: profileStore)
        self.fullscreenDetector = FullscreenDetector()
        self.eventTap = EventTapManager(triggerButton: profileStore.settings.triggerButton)
        self.executor = ActionExecutor(
            scriptOptions: ScriptExecutionOptions(timeout: profileStore.settings.scriptTimeoutSeconds)
        )

        let initial = contextEngine.currentProfile
        self.model = RingViewModel(
            geometry: RingGeometry(size: initial.ringSize, slotCount: initial.slotCount),
            slots: initial.renderedSlots(),
            showLabels: profileStore.settings.showLabels
        )
    }

    // MARK: Lifecycle

    /// Returns false if accessibility permission is missing (the tap won't arm).
    @discardableResult
    public func start() -> Bool {
        applySettings()

        contextEngine.onProfileChange = { [weak self] profile in
            self?.model.slots = profile.renderedSlots()
        }
        contextEngine.start()

        eventTap.onEvent = { [weak self] event in
            guard let self else { return .passEvent }
            return self.handle(event)
        }
        // Pre-create the overlay so the first ring open is just a (fast) show —
        // keeps the event-tap callback well under its timeout.
        _ = ensureWindow()
        return eventTap.enable()
    }

    public func stop() {
        eventTap.disable()
        contextEngine.stop()
        stopPolling()
        state = .closed
        previewMode = false
        window?.hide()
    }

    /// Re-read settings into the live subsystems (after the settings UI changes).
    public func applySettings() {
        let s = profileStore.settings
        eventTap.triggerButton = s.triggerButton
        model.showLabels = s.showLabels
    }

    // MARK: Core loop

    private func ensureWindow() -> RingWindow {
        if let window { return window }
        let w = RingWindow(model: model)
        window = w
        return w
    }

    private func handle(_ event: MouseTapEvent) -> EventTapAction {
        switch event.phase {
        case .down:
            return onTriggerDown()
        case .drag:
            // We rarely receive drags (the suppressed down stops the OS emitting
            // them); the poll tracks the cursor instead. Suppress any that do
            // arrive while the ring is up so stray button-drags don't reach the app.
            return state == .closed ? .passEvent : .suppress
        case .up:
            return onTriggerUp()
        }
    }

    private func onTriggerDown() -> EventTapAction {
        switch state {
        case .closed:
            // Master switch off, or a fullscreen game / suppressed app is frontmost
            // → let the click through untouched.
            if !profileStore.settings.enabled {
                expectingTriggerUp = false
                return .passEvent
            }
            if profileStore.settings.suppressInFullscreen, fullscreenDetector.shouldSuppressRing() {
                expectingTriggerUp = false
                return .passEvent
            }
            previewMode = false
            showRing(for: contextEngine.currentProfile, at: NSEvent.mouseLocation)
            state = .holding
            openedAt = Date()
            startPolling()
            expectingTriggerUp = true
            return .suppress   // swallow the press so the app's own binding doesn't fire

        case .sticky:
            // Second click confirms: fire the highlighted slot (cancel in center).
            let selected = window?.updateSelection(cursor: NSEvent.mouseLocation)
            finishRing(selecting: selected)
            expectingTriggerUp = true
            return .suppress

        case .holding:
            // A second down while still holding (e.g. a missed up) — resolve now.
            let selected = window?.updateSelection(cursor: NSEvent.mouseLocation)
            finishRing(selecting: selected)
            expectingTriggerUp = true
            return .suppress
        }
    }

    private func onTriggerUp() -> EventTapAction {
        let suppressUp = expectingTriggerUp
        expectingTriggerUp = false
        if state == .holding { resolveHoldRelease() }
        return suppressUp ? .suppress : .passEvent
    }

    /// Interpret a release while holding: fire a selection, or — for a quick tap
    /// with nothing selected — switch to a sticky (stays-open) ring.
    private func resolveHoldRelease() {
        let selected = window?.updateSelection(cursor: NSEvent.mouseLocation)
        if let selected {
            finishRing(selecting: selected)
            return
        }
        if Date().timeIntervalSince(openedAt) < tapThreshold {
            enterSticky()                  // a tap → keep the ring open
        } else {
            finishRing(selecting: nil)     // a deliberate hold in the center → cancel
        }
    }

    private func enterSticky() {
        state = .sticky
        let work = DispatchWorkItem { [weak self] in
            guard let self, self.state == .sticky else { return }
            self.finishRing(selecting: nil)
        }
        stickyTimeout = work
        DispatchQueue.main.asyncAfter(deadline: .now() + 6, execute: work)
    }

    private func showRing(for profile: RingProfile, at point: CGPoint) {
        let geometry = RingGeometry(size: profile.ringSize, slotCount: profile.slotCount)
        let w = ensureWindow()
        w.apply(geometry: geometry)
        model.slots = profile.renderedSlots()
        model.profileName = profile.name
        w.show(at: point)
    }

    /// Close the ring and fire `selected`'s action if present. Idempotent.
    private func finishRing(selecting selected: Int?) {
        guard state != .closed || previewMode else { return }
        state = .closed
        let wasPreview = previewMode
        previewMode = false
        stopPolling()
        window?.hide()

        guard
            !wasPreview, profileStore.settings.executeOnRelease,
            let index = selected,
            let slot = model.slots.first(where: { $0.position == index }),
            let action = slot.action
        else { return }

        Task { [weak self] in
            guard let self else { return }
            let result = await self.executor.execute(action)
            self.lastResult = result
            self.onActionExecuted?(slot, result)
        }
    }

    // MARK: Cursor polling (tracks the cursor; release comes from the up event)

    private func startPolling() {
        stopPolling()
        let timer = Timer(timeInterval: 1.0 / 120.0, repeats: true) { [weak self] _ in
            MainActor.assumeIsolated {
                guard let self, self.state != .closed || self.previewMode else { return }
                self.window?.updateSelection(cursor: NSEvent.mouseLocation)
            }
        }
        RunLoop.main.add(timer, forMode: .common)
        pollTimer = timer
    }

    private func stopPolling() {
        pollTimer?.invalidate(); pollTimer = nil
        previewDismiss?.cancel(); previewDismiss = nil
        stickyTimeout?.cancel(); stickyTimeout = nil
    }

    // MARK: Preview (verify the ring without granting Accessibility)

    /// Show the ring centered on the active screen, tracking the cursor, then
    /// auto-dismiss. Lets the user see the ring respond before wiring the trigger.
    public func showPreview(duration: TimeInterval = 4) {
        previewMode = true
        let screen = NSScreen.main ?? NSScreen.screens.first
        let point = screen.map { CGPoint(x: $0.frame.midX, y: $0.frame.midY) } ?? NSEvent.mouseLocation
        showRing(for: contextEngine.currentProfile, at: point)
        startPolling()   // cancels any prior previewDismiss
        let work = DispatchWorkItem { [weak self] in
            guard let self, self.previewMode else { return }
            self.finishRing(selecting: nil)
        }
        previewDismiss = work
        DispatchQueue.main.asyncAfter(deadline: .now() + duration, execute: work)
    }

    // MARK: Button recording (for settings)

    /// Capture the next mouse-button press as the new trigger button.
    public func recordTriggerButton(_ completion: @escaping (Int) -> Void) {
        eventTap.onButtonRecorded = { [weak self] button in
            guard let self else { return }
            self.profileStore.settings.triggerButton = button
            self.profileStore.saveSettings()
            self.eventTap.triggerButton = button
            completion(button)
        }
        eventTap.startRecording()
    }

    public func cancelRecording() { eventTap.stopRecording() }
}
#endif

File: windows/src/main/ringController.js
'use strict';
// RingController — port of Controller/RingController.swift. The orchestrator:
// trigger events → gesture state machine → overlay → executor.
//
// Constants are verbatim from the mac build:
//   tapThreshold 0.30s · sticky timeout 6s · preview 4s · poll 120Hz
//
// Gestures:
//   HOLD  — press, drag to a slot, release to fire; release in center cancels.
//   CLICK — quick tap (<0.30s, no slot) opens a sticky ring; second click
//           fires the highlighted slot (center = cancel); auto-cancel at 6s.
//
// expectingTriggerUp pairs suppression: an up is suppressed iff its down was,
// even if the ring closed in between — no dangling up ever reaches the app.

const { screen } = require('electron');
const { RingGeometry } = require('../shared/geometry');
const { renderedSlots } = require('../shared/models');
const { isFullscreenAppActive } = require('./win32/fullscreen');

const TAP_THRESHOLD_MS = 300;
const STICKY_TIMEOUT_MS = 6000;
const PREVIEW_DURATION_MS = 4000;
const POLL_MS = 1000 / 120;

const CLOSED = 'closed';
const HOLDING = 'holding';
const STICKY = 'sticky';

class RingController {
  /**
   * @param {object} deps { store, hook, overlay, contextEngine, executor }
   */
  constructor({ store, hook, overlay, contextEngine, executor }) {
    this.store = store;
    this.hook = hook;
    this.overlay = overlay;
    this.contextEngine = contextEngine;
    this.executor = executor;

    this.state = CLOSED;
    this.openedAt = 0;
    this.expectingTriggerUp = false;
    this.previewMode = false;
    this.pollTimer = null;
    this.previewDismiss = null;
    this.stickyTimeout = null;
    this.lastResult = null;
    this.onActionExecuted = null; // (slot, result) => void

    // The WH_MOUSE_LL hook proc must return its suppress/pass verdict FAST
    // (MSDN: exceed LowLevelHooksTimeout and Windows silently drops the hook).
    // So the fullscreen verdict is precomputed on a cheap timer and read from
    // this cache inside the callback — never queried synchronously there.
    this._fullscreenSuppressed = false;
    this._fullscreenTimer = null;
  }

  get isArmed() { return this.hook.mode !== 'none'; }

  /** @returns {string} the hook mode that ended up active */
  start() {
    this.applySettings();
    this.contextEngine.onProfileChange = (profile) => {
      // Live slot refresh while open (mac parity); geometry sticks until next open.
      if (this.state !== CLOSED || this.previewMode) {
        this.overlay.slots = renderedSlots(profile);
        this.overlay.profileName = profile.name;
        this.overlay._send(this.overlay._fullState());
      }
    };
    this.contextEngine.start();
    this.hook.onTriggerDown = (pt) => this._onTriggerDown(pt);
    this.hook.onTriggerUp = (pt) => this._onTriggerUp(pt);
    this.overlay.ensureWindow(); // pre-create so first open is a fast show
    this._startFullscreenWatch();
    return this.hook.start();
  }

  stop() {
    this.hook.stop();
    this.contextEngine.stop();
    this._stopPolling();
    if (this._fullscreenTimer) { clearInterval(this._fullscreenTimer); this._fullscreenTimer = null; }
    this.state = CLOSED;
    this.previewMode = false;
    this.overlay.hide();
  }

  _startFullscreenWatch() {
    if (this._fullscreenTimer) return;
    const tick = () => {
      try {
        this._fullscreenSuppressed =
          this.store.settings.suppressInFullscreen && isFullscreenAppActive();
      } catch { this._fullscreenSuppressed = false; }
    };
    tick();
    this._fullscreenTimer = setInterval(tick, 250);
  }

  applySettings() {
    const s = this.store.settings;
    this.hook.triggerButton = s.triggerButton;
    this.executor.timeoutSeconds = s.scriptTimeoutSeconds;
    // Propagate showLabels live so toggling it updates an open/next ring (parity).
    this.overlay.showLabels = s.showLabels;
    if (this.state !== CLOSED || this.previewMode) {
      this.overlay._send({ showLabels: s.showLabels });
    }
  }

  // ---- trigger handling (runs synchronously inside the hook callback) ---------

  _onTriggerDown() {
    const s = this.store.settings;
    if (this.state === CLOSED) {
      if (!s.enabled) { this.expectingTriggerUp = false; return 'pass'; }
      if (this._fullscreenSuppressed) { // cached — not queried in the hook proc
        this.expectingTriggerUp = false;
        return 'pass'; // let the game have the click
      }
      // Commit the synchronous state and verdict NOW; do the heavy work (screen
      // query + window show + poll start) on the next tick so the hook proc
      // returns immediately and never trips LowLevelHooksTimeout.
      this.previewMode = false;
      this.state = HOLDING;
      this.openedAt = Date.now();
      this.expectingTriggerUp = true;
      setImmediate(() => {
        if (this.state !== HOLDING) return; // gesture already resolved
        this._showRing(this.contextEngine.currentProfile, screen.getCursorScreenPoint());
        this._startPolling();
      });
      return 'suppress';
    }
    // STICKY second click confirms · HOLDING double-down = missed up: resolve now.
    this.expectingTriggerUp = true;
    setImmediate(() => {
      const selected = this.overlay.updateSelection(screen.getCursorScreenPoint());
      this._finishRing(selected);
    });
    return 'suppress';
  }

  _onTriggerUp() {
    const suppressUp = this.expectingTriggerUp;
    this.expectingTriggerUp = false;
    if (this.state === HOLDING) this._resolveHoldRelease();
    return suppressUp ? 'suppress' : 'pass';
  }

  _resolveHoldRelease() {
    const selected = this.overlay.updateSelection(screen.getCursorScreenPoint());
    if (selected != null) { this._finishRing(selected); return; }
    if (Date.now() - this.openedAt < TAP_THRESHOLD_MS) this._enterSticky();
    else this._finishRing(null);
  }

  _enterSticky() {
    this.state = STICKY;
    if (this.stickyTimeout) clearTimeout(this.stickyTimeout);
    this.stickyTimeout = setTimeout(() => {
      if (this.state === STICKY) this._finishRing(null);
    }, STICKY_TIMEOUT_MS);
  }

  // ---- show / hide -----------------------------------------------------------

  _showRing(profile, at) {
    const geometry = RingGeometry.forSize(profile.ringSize, profile.slotCount);
    this.overlay.apply(geometry);
    this.overlay.show({
      at,
      slots: renderedSlots(profile),
      profileName: profile.name,
      showLabels: this.store.settings.showLabels,
    });
  }

  /** Close + maybe execute. Idempotent. */
  _finishRing(selected) {
    if (this.state === CLOSED && !this.previewMode) return;
    this.state = CLOSED;
    const wasPreview = this.previewMode;
    this.previewMode = false;
    this._stopPolling();
    this.overlay.hide();

    if (wasPreview || !this.store.settings.executeOnRelease || selected == null) return;
    const slot = this.overlay.slots.find((sl) => sl.position === selected);
    if (!slot || slot.action == null) return;
    // Ring is already hidden — execution must not block the UI.
    this.executor.execute(slot.action).then((result) => {
      this.lastResult = result;
      if (!result.ok) console.error(`ActionRing: '${slot.label}' failed: ${result.error}`);
      this.onActionExecuted?.(slot, result);
    });
  }

  // ---- polling (release is NEVER inferred from the poll — only from up) --------

  _startPolling() {
    this._stopPolling();
    this.pollTimer = setInterval(() => {
      if (this.state === CLOSED && !this.previewMode) return;
      this.overlay.updateSelection(screen.getCursorScreenPoint());
    }, POLL_MS);
  }

  /** Stops the poll AND cancels preview/sticky timers (mac parity). */
  _stopPolling() {
    if (this.pollTimer) { clearInterval(this.pollTimer); this.pollTimer = null; }
    if (this.previewDismiss) { clearTimeout(this.previewDismiss); this.previewDismiss = null; }
    if (this.stickyTimeout) { clearTimeout(this.stickyTimeout); this.stickyTimeout = null; }
  }

  // ---- preview (state stays CLOSED; never executes) -----------------------------

  showPreview(durationMs = PREVIEW_DURATION_MS) {
    this.previewMode = true;
    const display = screen.getDisplayNearestPoint(screen.getCursorScreenPoint());
    const { x, y, width, height } = display.workArea;
    const center = { x: Math.round(x + width / 2), y: Math.round(y + height / 2) };
    this._showRing(this.contextEngine.currentProfile, center);
    this._startPolling(); // cancels any prior previewDismiss via _stopPolling
    this.previewDismiss = setTimeout(() => {
      if (this.previewMode) this._finishRing(null);
    }, durationMs);
  }

  // ---- trigger recording ---------------------------------------------------------

  recordTriggerButton(completion) {
    this.hook.onButtonRecorded = (button) => {
      this.store.settings.triggerButton = button;
      this.store.saveSettings();
      this.hook.triggerButton = button;
      completion?.(button);
    };
    this.hook.recording = true;
  }

  cancelRecording() { this.hook.recording = false; }
}

module.exports = {
  RingController,
  TAP_THRESHOLD_MS, STICKY_TIMEOUT_MS, PREVIEW_DURATION_MS, POLL_MS,
};
